driver/src/orderbook/shadow_orderbook.rs

```diff
@@ -0,0 +1,312 @@
+//! This module implements a shadowed orderbook, that is a main orderbook
+//! retrieval method that gets shadowed by a secondary one, and where the result
+//! is compared between the two.
+//!
+//! This is useful for validating alternate account retrieval methods during
+//! development.
+
+#![allow(dead_code)]
+
+use super::StableXOrderBookReading;
+use crate::models::{AccountState, Order, TokenId};
+use anyhow::Result;
+use ethcontract::Address;
+use std::collections::{HashMap, HashSet};
+
+/// A type definition representing a complete orderbook.
+type Orderbook = (AccountState, Vec<Order>);
+
+/// A struct representing a diffs in two queried orderbooks.
+#[derive(Debug, PartialEq)]
+struct Diff(Vec<BalanceChange>, Vec<OrderChange>);
+
+impl Diff {
+    /// Reads an orderbook with the specified reader at the given batch and
+    /// compares its results to the expected orderbook, returing a diff of the
+    /// two.
+    ///
+    /// The diff is expressed in changes in the primary orderbook relative to
+    /// the orderbook read with the reader.
+    fn compare_to_reader(
+        primary_orderbook: &Orderbook,
+        reader: &dyn StableXOrderBookReading,
+        batch_id: u32,
+    ) -> Result<Self> {
+        let shadow = reader.get_auction_data(batch_id.into())?;
```

Comment:
I'm wondering if it would be cleaner to have the `Diff` struct only take care of diffing and pass it in the orderbooks already fetched. Not a huge difference, but maybe it encapsulates responsibilities in a more straightforward way.

Reply:
Oh yeah! That makes perfect sense. I’ll change it.